build: integrate prepare source from composer

lzap opened the pull request:
I noticed this automated PR failed:

https://github.com/osbuild/image-builder/pull/1475

This is a chance to integrate changes from composer which uses a new Go feature "toolbox", hopefully it will prevent from automated failures.

lzap commented:
Hey @croissanne and @ezr-ondrej this needs an attention - other PRs are failing because of a version requirement, I believe it is great time to fix prepare sources for once.

There is a tread on `#osbuild` channel about this, this was merged into composer, then reverted by @achilleas-k and now the final version is heading in: https://github.com/osbuild/osbuild-composer/pull/4600 also being integrated elsewhere: https://github.com/osbuild/weldr-client/pull/155

lzap commented on tools/prepare-source.sh:
```diff
@@ -1,25 +1,29 @@
 
-GOBIN=$TOOLS_PATH $GO_BINARY generate -x -mod=mod ./...
+# Generate source (skip vendor/):
+GOBIN=$TOOLS_PATH go generate -x ./cmd/... ./internal/...
```
The `TOOLS_PATH` will disappear once we upgrade to 1.24 which features the all new Go tool pinning feature. In short, external tools will be added to `go.mod` now ensuring we all are on the same version. Until then, this is I believe a good workaround for the time being.

ezr-ondrej commented on tools/prepare-source.sh:
```diff
@@ -1,25 +1,32 @@
-$GO_BINARY download
+# Check latest Go version for the minor we're using
+LATEST=$(curl -s https://endoflife.date/api/go/"${GO_MAJOR_VER}".json  | jq -r .latest)
+if test "$LATEST" != "go$GO_VERSION"; then
```
This doesn't work, I see the WARNING even tho we are on latest, I believe the issue is:
```suggestion
if test "$LATEST" != "$GO_VERSION"; then
```

lzap replied:
Right, rebased.

ezr-ondrej reviewed (commented):
Works for me, one nit, apart of that it works... and if it works, merge it :stuck_out_tongue:

ezr-ondrej commented:
Sorry I wasn't fast enough and there is conflict now :/

lzap commented:
> Sorry I wasn't fast enough and there is conflict now :/

This is fine, the other patch was more important it fixed a CICD problem. Rebased.

ezr-ondrej commented:
The Checks CI fails with following error, I'm not sure what it's trying to tell me :)
```
jsonschema: "run" does not validate with "/properties/run/additionalProperties": additional properties 'skip-dirs' not allowed
jsonschema: "" does not validate with "/additionalProperties": additional properties 'exclude-dirs' not allowed
Failed executing command with error: the configuration contains invalid elements
```

lzap commented:
Fixed, this can be dropped from the linter config as `vendor/` was dropped from this repository.

Also integrating https://github.com/osbuild/image-builder-crc/pull/1484 update into this PR as well.

ezr-ondrej commented:
Hmm container test is trying to build an container and fails
`go: go.mod requires go >= 1.22.12 (running go 1.22.9; GOTOOLCHAIN=local)` go-toolset seems to be on 1.22.9 :/

lzap commented:
Yeah the warning I implemented is sort of useless since Go Toolset in RHEL is few versions behind usually. But we now have the working in composer to so let's stick with it? Rebased.

ezr-ondrej commented:
/retest

ezr-ondrej commented:
The failures feel unrelated :thinking:
Lets discuss this with osbuild QE tomorrow :)

lzap commented:
Ephemeral failure. /retest

tkoscieln commented:
@ezr-ondrej hello, this is my mistake and I know exactly what broke here. I accidentally forgot to push one commit before merging new changes to our plugin yesterday, @lzap the missing commit is a fix for the issues I had with what API returns that we discussed last week. Will fix this asap.

tkoscieln commented:
@ezr-ondrej @lzap sorry for the delay, was facing some errors from Quay during the IQE container push. Should be fixed now.
/retest

lzap commented:
/retest

lzap commented:
@ezr-ondrej @croissanne only konflux fails, can we merge this?

tkoscieln commented:
@lzap it should be okay to merge despite konflux failing because the fail for bonfire is sort of expected

ezr-ondrej commented:
Yes, Konflux merge is expected! :) nothing to do about it ATM definitely not caused by this PR.

ezr-ondrej reviewed (approved):
Works, LGTM :+1:
